misc/vim/plugin/nit.vim

```diff
@@ -80,3 +80,148 @@ function ForceNitComplete()
+		elseif name =~ '^'.a:base
+			" Common-prefix match
+			call NitOmnifuncAddAMatch(prefix_matches, words, name)
+		elseif name =~ a:base
+			" Substring match
+			call NitOmnifuncAddAMatch(substring_matches, words, name)
+		endif
+	endfor
+
+	" Assemble the final match list
+	call extend(a:matches, sort(prefix_matches))
+	call extend(a:matches, sort(substring_matches))
+endfun
+
+" Internal function to search parse the information from a metadata line
+fun NitOmnifuncAddAMatch(matches, words, name)
+	let pretty = get(a:words, 1, '')
+	let synopsis = get(a:words, 2, '')
+	let desc = get(a:words, 3, '')
+	let desc = join(split(desc, '#nnnn#', 1), "\n")
+	if strlen(pretty) > 40
+		let pretty = pretty[:37] . '...'
+	endif
+	call add(a:matches, {'word': a:name, 'abbr': pretty, 'menu': synopsis, 'info': desc, 'dup': 1})
+endfun
+
+" Omnifunc using metadata files generated by nitpick to offer better
+" contextual autocomplete for Nit source code.
+fun NitOmnifunc(findstart, base)
+	if a:findstart
+		" locate the start of the word
+		let line = getline('.')
+		let start = col('.') - 1
+		while start > 0 && line[start - 1] =~ '\w'
+			let start -= 1
+		endwhile
+		return start
+	else
+		" find keyword matching with "a:base"
+		let matches = []
+
+		" Advanced suggestions
+		let cursor_line = getline('.')
+
+		" Content of the line before the partial word
+		let line_prev_cursor = cursor_line[:col('.')-1]
+
+		" Char just before the partial word
+		let prev_char = line_prev_cursor[-1:]
+
+		" Nity words on the current line before the partial word
+		let prev_words = split(line_prev_cursor, '\W\+')
+
+		" The word right before the partial word
+		let prev_word = get(prev_words, -1, '')
+
+		" Have we found a promising heuristic yet?
+		let handled = 0
+
+		" Modules
+		if prev_word == 'import'
+			let handled = 1
+			call NitOmnifuncAddFromFile(a:base, matches, 'modules.txt')
+		endif
+
+		" Classes
+		if count(['new', 'super', 'class'], prev_word) > 0
+			let handled = 1
+			call NitOmnifuncAddFromFile(a:base, matches, 'classes.txt')
+		endif
+
+		" Properties
+		if prev_char == '.'
+			let handled = 1
+			call NitOmnifuncAddFromFile(a:base, matches, 'properties.txt')
+		endif
+
+		" If is nothing else...
+		if !handled
+			" it may be a keyword
+			if strlen(a:base) > 0
+				for keyword in ['module', 'import', 'class', 'abstract', 'interface',
+					\'universal', 'enum', 'end', 'fun', 'type', 'init', 'redef', 'is',
+					\'do', 'var', 'extern', 'public', 'protected', 'private', 'intrude',
+					\'if', 'then', 'else', 'while', 'loop', 'for', 'in', 'and', 'or',
+					\'not', 'implies', 'return', 'continue', 'break', 'abort', 'assert',
+					\'new', 'isa', 'once', 'super', 'self', 'true', 'false', 'null',
+					\'as', 'nullable', 'isset', 'label']
+
+					if keyword =~ '^' . a:base
+						call add(matches, keyword)
+					endif
+				endfor
+			endif
+
+			" it may still be a method call or property access
```

Comment:
If you have a partial word, you could discriminate with the first letter: lowercase means property, uppercase means class